lib/iex/lib/iex/introspection.ex

```diff
@@ -39,112 +30,183 @@ defmodule IEx.Introspection do
+    helpers = IEx.Config.doc_helpers(:helpers)
+    result = Enum.find_value(modules,
+               fn(module) ->
+                  Enum.find_value(helpers,
+                        fn(mod) -> found_help(mod, module, function) end)
+               end)
+    case result do
+      {:found, doc_list} -> display_doc_list(:found, doc_list)
+      _                  ->
+              print_error(result, "No documentation for #{to_string(function)} was found")
+    end
 
     dont_display_result
   end
 
   def h(module, function) when is_atom(module) and is_atom(function) do
-    case h_mod_fun(module, function) do
-      :ok ->
-        :ok
-      :no_docs ->
-        puts_error("#{inspect module} was not compiled with docs")
-      :not_found ->
-        nodocs("#{inspect module}.#{function}")
-    end
+    Enum.map(get_docs(module, function),
+      fn {status, doc_list} -> display_doc_list(status, doc_list) end)
 
     dont_display_result
   end
 
-  defp h_mod_fun(mod, fun) when is_atom(mod) do
-    if docs = Code.get_docs(mod, :docs) do
-      result = for {{f, arity}, _line, _type, _args, doc} <- docs, fun == f, doc != false do
-        h(mod, fun, arity)
-      end
-
-      if result != [], do: :ok, else: :not_found
-    else
-      :no_docs
-    end
-  end
-
   @doc """
   Prints the documentation for the given function
   and arity in the list of modules.
   """
   def h(modules, function, arity) when is_list(modules) and is_atom(function) and is_integer(arity) do
-    result =
-      Enum.find_value modules, false, fn module ->
-        h_mod_fun_arity(module, function, arity) == :ok
-      end
-
-    unless result, do: nodocs("#{function}/#{arity}")
+    helpers = IEx.Config.doc_helpers(:helpers)
+    result = Enum.find_value(modules,
+              fn(module) -> Enum.find_value(helpers,
+                             fn(mod) -> found_help(mod, module, function, arity) end)
+              end)
+
+    case result do
+      {:found, doc_list} -> display_doc_list(:found, doc_list)
+      _                  ->
+        print_error(result, "No documentation for #{to_string(function)}/#{inspect(arity)} was found" )
+    end
 
     dont_display_result
   end
 
   def h(module, function, arity) when is_atom(module) and is_atom(function) and is_integer(arity) do
-    case h_mod_fun_arity(module, function, arity) do
-      :ok ->
-        :ok
-      :no_docs ->
-        puts_error("#{inspect module} was not compiled with docs")
-      :not_found ->
-        nodocs("#{inspect module}.#{function}/#{arity}")
-    end
+    Enum.map(get_docs(module, function, arity),
+             fn {status, doc_list} -> display_doc_list(status, doc_list) end)
 
     dont_display_result
   end
 
-  defp h_mod_fun_arity(mod, fun, arity) when is_atom(mod) do
-    if docs = Code.get_docs(mod, :docs) do
-      doc = find_doc(docs, fun, arity, 4)
-            || find_default_doc(docs, fun, arity)
+  @doc """
+  Map over helpers and return a list of header, doc tuples.
```

Comment:
Docs should use the verbs in Present continuos, not in impreative form,
so it should read
> Maps over helpers and returns a list of header, doc tuples.